File: java/Old Raytracer Files/RayDemo03 - Multiple/Traceable.java
import java.util.*;
import java.awt.*;

/**
  An abstract class which serves as the ancestor of
  those classes which can be traced by the ray tracer.

  @version 1 June 2002
  @author Bob Crawford

*/

public abstract class Traceable
{
  String name = "";
  MatrixPair mm = new MatrixPair();  // Affine transformation for the object -
                                     // the modeling matrix.

  /**
    Converts a given world ray into the local coordinate system
    via the inverse of the modeling matrix.

    @param r The world ray
    @return The given ray expressed in the local coordinate system
  */
  public Ray getLocalRay(Ray r)
  {
    Vector3 worldOrigin = r.getOrigin();
    Vector3 worldPoint = r.pointAt(1);
    Vector3 localOrigin = mm.Inv.image(worldOrigin);
    Vector3 localPoint = mm.Inv.image(worldPoint);

    return new Ray(localOrigin, localPoint);
  }

  /**
    Returns an ordered list of the intersections of
    the current object with the given ray. If the
    ray does not intersect the object, null is returned.

    @param r The ray to intersect with the current object
    @return The ordered list of ray-object intersections;
            null if there are none
  */
  public abstract DoubleList intersections(Ray r);

  public abstract Color getColorAt(Vector3 p);

  public void dump()
  {
    System.out.println("Name: " + name);
  }
}

File: src/Tracer/Traceable.cs
using System.Drawing;
using System.Numerics;

namespace Tracer
{    
    public abstract class Traceable
    {
        public MatrixPair Model { get; private set; } = new MatrixPair();
        public string Name { get; set; }

        public Color BaseColor { get; set; } = Color.White;
        public Color SpecularColor { get; set; } = Color.White;
        public double SpecularExponent { get; set; } = 0;

        public Ray GetLocalRay(Ray ray)
        {
            var worldOrigin = ray.Origin;
            var worldPoint = ray.PointAt(1);
            var localOrigin = Model.Inverse.Image(worldOrigin);
            var localPoint = Model.Inverse.Image(worldPoint);
            
            return Ray.CreateFromPoints(localOrigin, localPoint);
        }

        public abstract float[] Intersections(Ray ray);
        public abstract Vector3 GetNormalAt(Vector3 p);
        public abstract Color GetBaseColorAt(Vector3 p);
    }
}